iodata/iodata.py

```diff
@@ -259,22 +260,31 @@ def atcorenums(self) -> np.ndarray:
             # https://github.com/PyCQA/pylint/issues/1694
             # pylint: disable=no-member
             result = self.atnums.astype(float)
-            self._atcorenums = result
+            self.atcorenums = result
         return result
 
     @atcorenums.setter
     def atcorenums(self, atcorenums):
         self._atcorenums = atcorenums
+        if self._charge is not None:
+            # _charge is treated as the dependent one, while atcorenums and
+            # nelec are treated as independent.
+            if self.nelec is None:
+                # Switch to storing _nelec.
+                self._nelec = atcorenums.sum() - self._charge
+            self._charge = None
 
     @property
     def charge(self) -> float:
         """Return the net charge of the system."""
-        if self.atcorenums is None:
+        # The internal _charge is used only if it cannot be derived.
+        if self.atcorenums is None or self.nelec is None:
             return self._charge
         return self.atcorenums.sum() - self.nelec
 
     @charge.setter
     def charge(self, charge: float):
+        # The internal _charge is used only if atcorenums is None.
         if self.atcorenums is None:
             self._charge = charge
         else:
```

Comment:
If data.charge is set to None, after setting data.atcorenums and data.charge, raise an error (Line 291):
```python
from iodata import IOData
d = IOData()
d.atnums = np.array([6, 1, 1, 1, 1])
d.charge = 1
d.charge = None
```
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/lmacaya/Desktop/horton3/fix_charge_issue/iodata/iodata/iodata.py", line 291, in charge
    self.nelec = self.atcorenums.sum() - charge
TypeError: unsupported operand type(s) for -: 'float' and 'NoneType'
```
I don't know if this is an issue, but I point it out just in case.

Reply:
Needs to be fixed. I guess this is comparable to one of @FarnazH 's comments. Thanks for bringing this up.

Reply:
This is also fixed now.